### ISSUE TITLE
"Restart" button for multiplayer games

### ISSUE THREAD

**Issue by netnazgul** [OpenRA/OpenRA#15181]
Singleplayer games have a "Restart" button, but not multiplayer ones. Sometimes it can be useful for multiplayer when one of the players is erroneously AFK or there is some misplay that players agree to re-play, but going the usual way requires to gather all the same players at a single server again, set all the same options and set teams properly. Obviously, it doesn't help all restart cases as the settings stay the same and the majority of restarts are due to bad lobby settings (explored map, game speed, etc). Maybe restarting the MP game can drop all the players into the same lobby with the same settings as before.

Another thing that will greatly benefit with the button is watching replays. As replays do not have any control over rewinding, rewatching some missed events requires a lot of overhead work to bring up the same replay again (and it gets worse the more replays there are). It all can be avoided easily with a single "Restart" button in the game menu.

**Comment by oraservers** [comment]
Like the idea for watching multiplayer replays. Do not like the idea for an ongoing multiplayer game unless this can be integrated with some player voting functionality

**Comment by dragunoff** [comment]
Partially a duplicate of #14104

**Comment by matjaeck** [comment]
> Partially a duplicate of #14104

This issue should focus on the idea of restarting multiplayer games and leave the restart replay part to the other issue referenced above.

**Comment by knarrff** [comment]
One question that came to my mind: why is there a restart button necessary in the first place? I was always a bit surprised when the end of a game (even a regular end, but just as well an early abort) does not throw me back into the lobby, but completely off the server. Couldn't this be implemented such that any end of a game does not throw players back into the server list, but instead back into the lobby, preferably with settings kept/saved and restored from the just played game?

While this would remove the obstacle of a voting system, it might mean to add something like a player status in the lobby that is 'currently still playing' (but the spot in the lobby is reserved), because not all players will leave at the same time. This kind-of replaces the voting system: a new game can only be started if everyone is indeed back in the lobby and new players can only enter the lobby if no player is still playing.

One possible objection I can foresee is "What happens when a few players decide to continue the game while others already left? Wouldn't those that left the game be left hanging in the lobby?" Yes, they would. They decided to drop out. They are still free to leave the server. At the moment, the moment they dropped out they would leave the server just as well.

**Comment by netnazgul** [comment]
@knarrff you can't get back into the same server lobby after exiting the game because that server is still occupied by other players/spectators and the game is "ongoing", unless you're the last one to exit. In the latter case it's still a race condition because server needs some time to loop restart and be updated by the master server.

There is no server lobby possible in parallel while server is active with a game.

**Comment by knarrff** [comment]
@netnazgul I know it's not possible at the moment. I don't see why it couldn't be made possible. Are there large technical difficulties?

**Comment by pchote** [comment]
It is not technically possible for the same server to have two different groups of people in two different states (some in the previous game world, others in the next lobby).

I would personally also like to see OpenRA's end-game behaviour fixed to automatically show a normal RTS-style score summary and move everybody back to the same lobby at the end of the game. Unfortunately there is strong vocal opposition from the multiplayer community, who prefer the way things are now.
